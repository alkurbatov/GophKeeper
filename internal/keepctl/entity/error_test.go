package entity_test

import (
	"fmt"
	"testing"

	"github.com/alkurbatov/goph-keeper/internal/keepctl/entity"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/require"
	"google.golang.org/genproto/googleapis/rpc/errdetails"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

func TestRequestErrorFromBasicError(t *testing.T) {
	sat := entity.NewRequestError(grpc.ErrServerStopped)

	snaps.MatchSnapshot(t, sat.Error())
}

func TestRequestErrorFromGenericGRPCError(t *testing.T) {
	gRPCErr := status.Errorf(codes.Unauthenticated, "bad login or password")
	sat := entity.NewRequestError(gRPCErr)

	snaps.MatchSnapshot(t, sat.Error())
}

func TestRequestErrorFromGRPCBadRequest(t *testing.T) {
	details := &errdetails.BadRequest{}
	details.FieldViolations = append(
		details.FieldViolations,
		&errdetails.BadRequest_FieldViolation{
			Field:       "username",
			Description: "not set",
		},
		&errdetails.BadRequest_FieldViolation{
			Field:       "security_key",
			Description: "not set",
		},
	)

	st := status.New(codes.InvalidArgument, "invalid request")
	st, err := st.WithDetails(details)
	require.NoError(t, err)

	sat := entity.NewRequestError(st.Err())

	snaps.MatchSnapshot(t, sat.Error())
}

func TestUnwrap(t *testing.T) {
	tt := []struct {
		name string
		err  error
	}{
		{
			name: "Test unwrap of RequestError",
			err:  entity.NewRequestError(grpc.ErrServerStopped),
		},
		{
			name: "Test unwrap of other errors",
			err:  grpc.ErrServerStopped,
		},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			err := fmt.Errorf("ErrorTest - TestUnwrap - SomeError: %w", tc.err)

			snaps.MatchSnapshot(t, entity.Unwrap(err).Error())
		})
	}
}
