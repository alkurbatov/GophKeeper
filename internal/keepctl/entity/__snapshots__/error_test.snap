
[TestRequestErrorFromBasicError - 1]
grpc: the server has been stopped (255)
---

[TestRequestErrorFromGRPCError - 1]
bad login or password (16)
---

[TestUnwrap/Test_unwrap_of_RequestError - 1]
grpc: the server has been stopped (255)
---

[TestUnwrap/Test_unwrap_of_other_errors - 1]
ErrorTest - TestUnwrap - SomeError: grpc: the server has been stopped
---

[TestRequestErrorFromGenericGRPCError - 1]
bad login or password (16)
---

[TestRequestErrorFromGRPCBadRequest - 1]
invalid request (3)
    "username": not set
    "security_key": not set
---
